lib/open_food_network/scope_variants_for_search.rb

```diff
@@ -68,11 +71,40 @@ def scope_to_eligible_for_subscriptions_in_distributor
       scope_variants_to_distributor(@variants, distributor)
     end
 
+    def scope_to_in_stock_only
+      @variants = @variants.where(variant_in_stock_sql)
+    end
+
     def scope_variants_to_distributor(variants, distributor)
       scoper = OpenFoodNetwork::ScopeVariantToHub.new(distributor)
       # Perform scoping after all filtering is done.
       # Filtering could be a problem on scoped variants.
       variants.each { |v| scoper.scope(v) }
     end
+
+    def variant_in_stock_sql
+      [
+        variant_override_in_stock_exists_sql,
+        spree_stock_item_in_stock_exists_sql
+      ].reject(&:blank?).join(" OR ")
+    end
+
+    def spree_stock_item_in_stock_exists_sql
+      return if distributor.present?
+
+      "EXISTS(
+         SELECT 1
+         FROM spree_stock_items
+         WHERE variant_id = spree_variants.id AND (backorderable = true OR count_on_hand > 0)
+      )"
+    end
+
+    def variant_override_in_stock_exists_sql
+      sql = "SELECT 1
+             FROM variant_overrides
+             WHERE variant_id = spree_variants.id AND (on_demand = true OR count_on_hand > 0)"
+      sql += " AND hub_id = #{distributor.id}" if distributor.present?
+      "EXISTS (#{sql})"
+    end
```

Comment:
This is complicated. I understand this as:

- If no distributor is selected then we look at all variants which are in stock at the producer or in stock for any distributor.
- If a distributor is selected then we only look at variants that have a variant override with stock present.

Is that how you understand it as well?

The form to create a new order from the backoffice asks for a distributor in the first step. So I'm wondering if we can simplify this and ignore the case when no distributor is present (return nil).

So let's look closer at the case that a distributor is present. If we have a variant override for that distributor and it has stock, the variant will appear. If there is no variant override and the producer stock is present then the variant will appear as well. So far so good.

But If there is a variant and the cound_on_hand is zero then the variant override SQL returns false and it looks at the producer stock. That is not correct. We should only look at the producer stock if there is no variant override or the variant override doesn't have a stock level set. It can be NULL which means that we still use the stock level of the producer.

Does that make sense?

Reply:
@mkllnk Yes it's complicated :sweat_smile: so simplifying this sounds good. I have changed it now so it always expects a distributor to be present, I didn't rebase yet but I can do later. How does this look now? Thanks again for the help.